File: Chap6-GraphicsViewFramework/6-8ResizableItemDemo/resizablerectitem.py
from PySide6.QtWidgets import QGraphicsRectItem, QGraphicsItem
from PySide6.QtCore import QRectF, QPointF, QSize
from PySide6.QtGui import QPen, QBrush, QPainter, QColor

class ResizableRectItem(QGraphicsRectItem):
    """
    A rectangle item that can be resized using handles on its corners
    """
    def __init__(self):
        super().__init__()
        self.handles_added_to_scene = False
        self.handle_list = []
        
        # Initialize handle rects
        self.topleft_handle_rect = QRectF()
        self.topright_handle_rect = QRectF()
        self.bottomright_handle_rect = QRectF()
        self.bottomleft_handle_rect = QRectF()
    
    def selectorFrameBounds(self):
        """
        Get the bounds of the selector frame
        
        Returns:
            QRectF: The bounds rectangle
        """
        return self.rect()
    
    def setSelectorFrameBounds(self, bounds_rect):
        """
        Set the bounds of the selector frame
        
        Args:
            bounds_rect: QRectF with the new bounds
        """
        self.prepareGeometryChange()
        self.setRect(bounds_rect)
        self.update()
    
    def boundingRect(self):
        """
        Get the bounding rectangle of the item
        
        Returns:
            QRectF: The bounding rectangle
        """
        return self.selectorFrameBounds()
    
    def paint(self, painter, option, widget):
        """
        Paint the item
        
        Args:
            painter: QPainter
            option: QStyleOptionGraphicsItem
            widget: QWidget
        """
        painter.save()
        painter.setBrush(self.brush())
        painter.drawRect(self.rect())
        self.drawHandlesIfNecessary()
        painter.restore()
    
    def drawHandlesIfNecessary(self):
        """
        Draw handles if the item is selected, remove them otherwise
        """
        if self.isSelected():
            self.drawHandles()
            self.handles_added_to_scene = True
        else:
            # Remove the handles
            for handle_item in self.handle_list:
                self.scene().removeItem(handle_item)
            
            # Clear the list after removing handles
            self.handle_list.clear()
            self.handles_added_to_scene = False
    
    def drawHandles(self):
        """
        Draw the handles for resizing
        """
        from handleitem import HandleItem
        
        # Populate handles in list if empty
        if len(self.handle_list) == 0:
            self.handle_list.append(HandleItem(HandleItem.TopLeft))
            self.handle_list.append(HandleItem(HandleItem.TopRight))
            self.handle_list.append(HandleItem(HandleItem.BottomRight))
            self.handle_list.append(HandleItem(HandleItem.BottomLeft))
        
        # Set up pen
        pen = QPen()
        pen.setWidth(2)
        pen.setColor(QColor(0, 0, 0))  # Qt::black
        
        # Top left handle
        top_left_corner = self.boundingRect().topLeft() + QPointF(-12.0, -12.0)
        self.topleft_handle_rect = QRectF(top_left_corner, QSize(12, 12))
        self.handle_list[0].setRect(self.topleft_handle_rect)
        
        if len(self.handle_list) > 0 and not self.handles_added_to_scene:
            self.handle_list[0].setPen(pen)
            self.handle_list[0].setBrush(QBrush(QColor(128, 128, 128)))  # Qt::gray
            self.scene().addItem(self.handle_list[0])
            self.handle_list[0].setParentItem(self)
        
        # Top right handle
        top_right_corner = self.boundingRect().topRight() + QPointF(0, -12.0)
        self.topright_handle_rect = QRectF(top_right_corner, QSize(12, 12))
        self.handle_list[1].setRect(self.topright_handle_rect)
        
        if len(self.handle_list) > 0 and not self.handles_added_to_scene:
            self.handle_list[1].setPen(pen)
            self.handle_list[1].setBrush(QBrush(QColor(128, 128, 128)))  # Qt::gray
            self.scene().addItem(self.handle_list[1])
            self.handle_list[1].setParentItem(self)
        
        # Bottom right handle
        bottom_right_corner = self.boundingRect().bottomRight() + QPointF(0, 0)
        self.bottomright_handle_rect = QRectF(bottom_right_corner, QSize(12, 12))
        self.handle_list[2].setRect(self.bottomright_handle_rect)
        
        if len(self.handle_list) > 0 and not self.handles_added_to_scene:
            self.handle_list[2].setPen(pen)
            self.handle_list[2].setBrush(QBrush(QColor(128, 128, 128)))  # Qt::gray
            self.scene().addItem(self.handle_list[2])
            self.handle_list[2].setParentItem(self)
        
        # Bottom left handle
        bottom_left_corner = self.boundingRect().bottomLeft() + QPointF(-12, 0)
        self.bottomleft_handle_rect = QRectF(bottom_left_corner, QSize(12, 12))
        self.handle_list[3].setRect(self.bottomleft_handle_rect)
        
        if len(self.handle_list) > 0 and not self.handles_added_to_scene:
            self.handle_list[3].setPen(pen)
            self.handle_list[3].setBrush(QBrush(QColor(128, 128, 128)))  # Qt::gray
            self.scene().addItem(self.handle_list[3])
            self.handle_list[3].setParentItem(self)
        
        self.handles_added_to_scene = True

File: 16-GraphicsViewFramework/08_ResizableItemDemo/resizablerectitem.cpp
#include "resizablerectitem.h"
#include <QPen>
#include <QGraphicsScene>
#include <QPainter>

ResizableRectItem::ResizableRectItem() : handlesAddedToScene(false)
{

}

QRectF ResizableRectItem::selectorFrameBounds() const
{
    return rect();
}

void ResizableRectItem::setSelectorFrameBounds(QRectF boundsRect)
{

    prepareGeometryChange();
    setRect(boundsRect);
    update();

}

QRectF ResizableRectItem::boundingRect() const
{
    return selectorFrameBounds();
}

void ResizableRectItem::paint(QPainter *painter, const QStyleOptionGraphicsItem *option,
                              QWidget *widget)
{
    Q_UNUSED(option);
    Q_UNUSED(widget);
    painter->save();

    painter->setBrush(brush());

    painter->drawRect(rect());

    drawHandlesIfNecessary();

    painter->restore();
}

void ResizableRectItem::drawHandlesIfNecessary()
{
    if(isSelected()){
        drawHandles();
        handlesAddedToScene = true;
    }else{
        //Remove the handles
        foreach (HandleItem * handleItem, handleList) {
            scene()->removeItem(handleItem);
        }
        qDeleteAll(handleList);
        handleList.clear();
        handlesAddedToScene = false;
    }

}

void ResizableRectItem::drawHandles()
{
    //Populate handles in list
    if(handleList.count() == 0){
        handleList.append(new HandleItem(HandleItem::TopLeft));
        handleList.append(new HandleItem(HandleItem::TopRight));
        handleList.append(new HandleItem(HandleItem::BottomRight));
        handleList.append(new HandleItem(HandleItem::BottomLeft));
    }


    //Set up pen
    QPen mPen;
    mPen.setWidth(2);
    mPen.setColor(Qt::black);

    //Top left handle
    QPointF topLeftCorner = boundingRect().topLeft() + QPointF(-12.0,-12.0);
    topleftHandleRect  = QRectF(topLeftCorner,QSize(12,12));
    handleList[0]->setRect(topleftHandleRect);
    if(!handleList.isEmpty() && (!handlesAddedToScene)){
        handleList[0]->setPen(mPen);
        handleList[0]->setBrush(QBrush(Qt::gray));
        scene()->addItem(handleList[0]);
        handleList[0]->setParentItem(this);

    }

    //Top right
    QPointF topRightCorner = boundingRect().topRight() + QPointF(0,-12.0);
    topRightHandleRect  = QRectF(topRightCorner,QSize(12,12));
    handleList[1]->setRect(topRightHandleRect);
    if(!handleList.isEmpty() && (!handlesAddedToScene)){
        handleList[1]->setPen(mPen);
        handleList[1]->setBrush(QBrush(Qt::gray));
        scene()->addItem(handleList[1]);
        handleList[1]->setParentItem(this);
    }

    //Bottom right
    QPointF bottomRightCorner = boundingRect().bottomRight() + QPointF(0,0);
    bottomRightHandleRect  = QRectF(bottomRightCorner,QSize(12,12));
    handleList[2]->setRect(bottomRightHandleRect);
    if(!handleList.isEmpty() && (!handlesAddedToScene)){
        handleList[2]->setPen(mPen);
        handleList[2]->setBrush(QBrush(Qt::gray));
        scene()->addItem(handleList[2]);
        handleList[2]->setParentItem(this);
    }

    //Bottom left
    QPointF bottomLeftCorner = boundingRect().bottomLeft() + QPointF(-12,0);
    bottomLeftHandleRect  = QRectF(bottomLeftCorner,QSize(12,12));
    handleList[3]->setRect(bottomLeftHandleRect);
    if(!handleList.isEmpty() && (!handlesAddedToScene)){
        handleList[3]->setPen(mPen);
        handleList[3]->setBrush(QBrush(Qt::gray));
        scene()->addItem(handleList[3]);
        handleList[3]->setParentItem(this);
    }

    handlesAddedToScene = true;

}
